core/templates/hash_map2.h

```diff
@@ -0,0 +1,530 @@
+
+#include "core/math/math_funcs.h"
+#include "core/os/memory.h"
+#include "core/templates/hashfuncs.h"
+#include "core/templates/paged_allocator.h"
+#include "core/templates/pair.h"
+/**
+ * A HashMap implementation that uses open addressing with Robin Hood hashing.
+ * Robin Hood hashing swaps out entries that have a smaller probing distance
+ * than the to-be-inserted entry, that evens out the average probing distance
+ * and enables faster lookups. Backward shift deletion is employed to further
+ * improve the performance and to avoid infinite loops in rare cases.
+ *
+ * Keys and values are stored in a double linked list by insertion order. This
+ * has a slight performance overhead on lookup, which can be mostly compensated
+ * using a paged allocator if required.
+ *
+ * The assignment operator copy the pairs from one map to the other.
+ */
+
+template <class TKey, class TValue>
+struct HashMap2Element {
+	HashMap2Element *next = nullptr;
+	HashMap2Element *prev = nullptr;
+	KeyValue<TKey, TValue> data;
+	HashMap2Element() {}
+	HashMap2Element(const TKey &p_key, const TValue &p_value) :
+			data(p_key, p_value) {}
+};
+
+template <class TKey, class TValue,
+		class Hasher = HashMapHasherDefault,
+		class Comparator = HashMapComparatorDefault<TKey>,
+		class Allocator = DefaultTypedAllocator<HashMap2Element<TKey, TValue>>>
+class HashMap2 {
+private:
+	Allocator element_alloc;
+	HashMap2Element<TKey, TValue> **elements = nullptr;
+	uint32_t *hashes = nullptr;
+	HashMap2Element<TKey, TValue> *head_element = nullptr;
+	HashMap2Element<TKey, TValue> *tail_element = nullptr;
+
+	uint32_t capacity = 0;
+
+	uint32_t num_elements = 0;
+
+	static const uint32_t EMPTY_HASH = 0;
+
+	_FORCE_INLINE_ uint32_t _hash(const TKey &p_key) const {
+		uint32_t hash = Hasher::hash(p_key);
+
+		if (unlikely(hash == EMPTY_HASH)) {
+			hash = EMPTY_HASH + 1;
+		}
+
+		return hash;
+	}
+
+	_FORCE_INLINE_ uint32_t _get_probe_length(uint32_t p_pos, uint32_t p_hash) const {
+		uint32_t original_pos = p_hash % capacity;
+		return (p_pos - original_pos + capacity) % capacity;
+	}
+
+	bool _lookup_pos(const TKey &p_key, uint32_t &r_pos) const {
+		if (elements == nullptr) {
+			return false; // Failed lookups, no elements
+		}
+
+		uint32_t hash = _hash(p_key);
+		uint32_t pos = hash % capacity;
+		uint32_t distance = 0;
+
+		while (true) {
+			if (hashes[pos] == EMPTY_HASH) {
+				return false;
+			}
+
+			if (distance > _get_probe_length(pos, hashes[pos])) {
+				return false;
+			}
+
+			if (hashes[pos] == hash && Comparator::compare(elements[pos]->data.key, p_key)) {
+				r_pos = pos;
+				return true;
+			}
+
+			pos = (pos + 1) % capacity;
+			distance++;
+		}
+	}
+
+	void _insert_with_hash(uint32_t p_hash, HashMap2Element<TKey, TValue> *p_value) {
+		uint32_t hash = p_hash;
+		uint32_t distance = 0;
+		uint32_t pos = hash % capacity;
+
+		while (true) {
+			if (hashes[pos] == EMPTY_HASH) {
+				elements[pos] = p_value;
+				hashes[pos] = p_hash;
```

Comment:
The swapped out hash is put into `hash`, so this should be using `hash` instead of `p_hash` I think.

Maybe the current value should have its own variable `value` too, so it's easier to spot something goes wrong.

Reply:
good point, fixed,thanks!